File: Breadth-first_search(BFS)/python/06_특정거리의도시찾기.py
# 이것이 취업을 위한 코딩테스트다 339p
# https://www.acmicpc.net/problem/18352
# 2

from collections import deque
import sys

def bfs(n):
    q = deque([n])
    distance[n] = 0

    while q:
        px = q.popleft()
        for nx in matrix[px]:
            if distance[nx] == -1:
                q.append(nx)
                distance[nx] = distance[px] + 1

N, M, K, X = map(int, sys.stdin.readline().split())
matrix = [[] for i in range(N + 1)]
distance = [-1] *(N + 1)

for _ in range(M):
    x, y = map(int, sys.stdin.readline().split())
    matrix[x].append(y)

bfs(X)

chk = False
for i in range(N + 1):
    if distance[i] == K:
        chk = True
        print(i)

if not(chk):
    print(-1)

File: Breadth-first_search(BFS)/c++/06_특정거리의_도시찾기.cpp
// 이것이 취업을 위한 코딩테스트다 339p
// https://www.acmicpc.net/problem/18352
// 2

#include <iostream>
#include <vector>
#include <queue>

using namespace std;

vector<int> matrix[300001];
vector<int> dist(300001, -1);
int N, M, K, X;

void bfs(int x)
{
	queue<int> q;
	q.push(x);
	dist[x] = 0;

	while (!q.empty()) {
		int px = q.front();
		q.pop();

		for (int i = 0; i < matrix[px].size(); i++) {
			int nx = matrix[px][i];
			if (dist[nx] == -1) {
				q.push(nx);
				dist[nx] = dist[px] + 1;
			}
		}
	}
}

int main()
{
	cin >> N >> M >> K >> X;
	for (int i = 0; i < M; i++) {
		int x, y;
		scanf("%d %d", &x, &y);
		matrix[x].push_back(y);
	}

	bfs(X);

	bool chk = false;
	for (int i = 0; i < N + 1; i++) {
		if (dist[i] == K) {
			printf("%d\n", i);
			chk = true;
		}
	}
	if (!chk)
		printf("-1\n");
}